components/browser/menu/src/main/java/mozilla/components/browser/menu/view/ExpandableLayout.kt

```diff
@@ -0,0 +1,379 @@
+            expandedHeight = minOf(expandedHeight, parentHeight)
+
+            // Ensure the collapsedHeight we calculated is not bigger than the expanded height
+            // now capped by parent height.
+            // This might happen if the menu is shown in landscape and there is no space to show
+            // the lastVisibleItemIndexWhenCollapsed.
+            if (collapsedHeight >= expandedHeight) {
+                // If there's no space to show the lastVisibleItemIndexWhenCollapsed even if the
+                // wrappedView is collapsed there's no need to collapse the view.
+                collapsedHeight = expandedHeight
+                isExpandInProgress = false
+                isCollapsed = false
+            }
+        }
+
+        return expandedHeight
+    }
+
+    @Suppress("WrongCall")
+    @VisibleForTesting
+    // Used for testing protected super.onMeasure(..) calls will be executed.
+    internal fun callParentOnMeasure(widthSpec: Int, heightSpec: Int) {
+        super.onMeasure(widthSpec, heightSpec)
+    }
+
+    @Suppress("WrongCall")
+    @VisibleForTesting
+    // Used for testing protected super.onInterceptTouchEvent(..) calls will be executed.
+    internal fun callParentOnInterceptTouchEvent(ev: MotionEvent?): Boolean {
+        return super.onInterceptTouchEvent(ev)
+    }
+
+    /**
+     * Whether based on the previous movements, when considering this [event]
+     * it can be inferred that the user is currently scrolling up.
+     */
+    @VisibleForTesting
+    internal fun isScrollingUp(event: MotionEvent): Boolean {
+        val yDistance = initialYCoord - event.y
+
+        return yDistance >= touchSlop
+    }
+
+    // We need a dynamic way to get the intended collapsed height of this view before it will be laid out on the screen.
+    // This method assumes the following layout:
+    //          ____________________________________________________
+    //  this -> |              ----------------------------------- |
+    //          | ViewGroup -> |                ---------------- | |
+    //          |              | RecyclerView-> |     View     | | |
+    //          |              |                |     View     | | |
+    //          |              |                |     View     | | |
+    //          |              |                |  SpecialView | | |
+    //          |              |                |     View     | | |
+    //          |              |                ---------------- | |
+    //          |              ----------------------------------- |
+    //          ----------------------------------------------------
+    // for which we want to measure the distance (height) between [this#top, half of SpecialView].
+    // That distance will be the collapsed height of this View used when this will be first shown on the screen.
+    // Users will be able to afterwards expand this View to the full height.
```

Comment:
Fantastic 👏

Reply:
Hmm. Thanks for pointing this.
Looking at it now I think using `the ViewGroup` instead of `this View` in the last comment lines would be desired.